enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,199 @@
+---
+title: loadbalancer-service-support
+authors:
+  - "@pliurh"
+reviewers:
+  - "@copejon, MicroShift contributor"
+  - "@fzdarsky, MicroShift architect"
+  - "@ggiguash, MicroShift contributor"
+  - "@dhellmann, MicroShift contributor"
+  - "@oglok, MicroShift contributor"
+  - "@zshi-redhat, MicroShift "
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - None
+creation-date: 2023-01-13
+last-updated: 2023-01-13
+tracking-link:
+  - https://issues.redhat.com/browse/NP-604
+---
+
+# MicroShift Service of Loadbalancer Type Support
+
+## Summary
+
+Customers using loadbalancer type Service, doesn't want to make changes to their
+workloads/deployment for MicroShift as they use other k8s flavors as well. It is
+also promised by us that users do not have to modify workload manifests that
+work on OpenShift to run on MicroShift (as long as they do not use APIs
+unsupported by MicroShift).
+
+## Motivation
+
+Normally, the loadbalancer type Service is supported by a Kubernetes cluster
+with external load balancers. In Microshift, as there is only one node in the
+cluster, we don't really need to have a load balancer. However, from user
+experience perspective, it's important for customers to be able to migrate
+the workload to Microshift without modification. So we need to find a way to
+support the loadbalancer type Service without a load balancer.
+
+### User Stories
+
+As a Microshift user, I want to reuse the Service manifest that with type of
+`LoadBalancer` without modification. I don't want to allocate a new IP address
+to the Node as the loadbalancer VIP. I want to use the NodeIP as the ingress IP
+of the loadbalancer Service. 
+
+### Goals
+
+* Allow users to deploy the load balancer type Service to Microshift.
+* The node IP shall be used as the ingress IP of the Service.
+* As we will support pluggable CNI in Microshift. The implementation shall be
+also compatible with CNIs other than OVN-Kubernetes.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+To support the loadbalancer type Service, we need to create a new controller
+component that watch the Service objects. It shall be responsible for:
+
+1. work with CNIs to plumb the ingress traffic to the endpoint pods.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+### Workflow Description
+
+**cluster user** is a human user responsible for provisioning Services to a
+Microshift cluster.
+
+1. The cluster user deploys a Service of type: LoadBalancer 
+2. The loadbalancer service controller watches the creation of the service. It
+will update the Service's `.status.loadBalancer` field, and help the CNIs to 
+plumbing the ingress traffic to the endpoint pods.
```

Comment:
does it help to reach the endpoint pod directly or to the backed clusterIP service?

Reply:
It depends on which CNI we are talking about. For OVN-K, it doesn't need to do the plumbing work. OVN-K would do that. For other CNIs, if we take the k3s approach. It will create a pod on a node, and the pod will add iptables DNAT rules steering the traffic to the nodePort of the service.

Reply:
How/where does OVN-k do the plumbing?

Reply:
An alternative to what klipper does could be to simply setup DNAT rules on the system from microshift (by first checking the port is free in the system). In this way we avoid cost of additional pods on the system, and we also keep some control over the need to re-install the rules (i.e. firewalld reload)

An advantage to this could be that.. if those rules are compatible with whatever OVNk does under the hood we could serve the purpose of both OVNk and potentially other plugins.

Reply:
> How/where does OVN-k do the plumbing?

OVN-K watches the status of the service. It will add iptables rules that DNAT the ingress traffic to the clusterIP of the service.

Yes, I think we can also use microshift to insert the iptables rules instead of using klipper.